javalib/src/main/scala/java/util/Formatter.scala

```diff
@@ -79,1587 +91,989 @@ class Formatter(
 
-    private var dateTimeUtil: DateTimeUtil = _
+      val arg = if (conversion == '%' || conversion == 'n') {
+        /* No argument. Make sure not to bump `lastImplicitArgIndex` nor to
+         * affect `lastArgIndex`.
+         */
+        null
+      } else {
 
-    private def getNumberFormat(): NumberFormat = {
-      if (null == numberFormat)
-        numberFormat = NumberFormat.getInstance(locale)
-      numberFormat
-    }
+        val argIndex = if (flags.useLastIndex) {
+          // Explicitly use the last index
+          lastArgIndex
+        } else {
+          val i = parsePositiveIntSilent(optGroup(1), default = 0)
+          if (i == 0) {
+            // Either there is no explicit index, or the explicit index is 0
+            lastImplicitArgIndex += 1
+            lastImplicitArgIndex
+          } else if (i < 0) {
+            // Cannot be parsed, same as useLastIndex
+            lastArgIndex
+          } else {
+            // Could be parsed, this is the index
+            i
+          }
+        }
 
-    private def getDecimalFormatSymbols(): DecimalFormatSymbols = {
-      if (null == decimalFormatSymbols)
-        decimalFormatSymbols = new DecimalFormatSymbols(locale)
-      decimalFormatSymbols
-    }
+        val conversionStr = conversion.toString
+        if ("bBhHsScCdoxXeEgGfn%".indexOf(conversionStr) < 0)
+          throw new UnknownFormatConversionException(conversionStr)
 
-    def transform(token: FormatToken, argument: Object): String = {
-      this.formatToken = token
-      this.arg = argument
-
-      var result =
-        token.getConversionType() match {
-          case 'B' | 'b' => transformFromBoolean()
-          case 'H' | 'h' => transformFromHashCode()
-          case 'S' | 's' => transformFromString()
-          case 'C' | 'c' => transformFromCharacter()
-          case 'd' | 'o' | 'x' | 'X' =>
-            if (null == arg || arg.isInstanceOf[BigInteger])
-              transformFromBigInteger()
-            else
-              transformFromInteger()
-          case 'e' | 'E' | 'g' | 'G' | 'f' | 'a' | 'A' =>
-            transformFromFloat()
-          case '%' => transformFromPercent()
-          case 'n' => transformFromLineSeparator()
-          case 't' => transformFromDateTime()
-          case unknown =>
-            throw new UnknownFormatConversionException(String.valueOf(unknown))
+        if (argIndex <= 0 || argIndex > args.length) {
+          throw new MissingFormatArgumentException("%" + matcher.group())
         }
 
-      if (Character.isUpperCase(token.getConversionType())) {
-        if (null != result) {
-          // Porting note: Harmony does this but this.locale should be respected
-          result = result.toUpperCase(Locale.US)
+        if (width < 0 &&
+            (isNumericConversion(conversion) && flags.zeroPad ||
+            flags.leftAlign)) {
+          throw new MissingFormatWidthException("%" + matcher.group())
         }
```

Comment:
Is there a reason that this is done differently than in Scala.js?

Reply:
Yes, the first one is complience with JVM, eg. `_.format("%0", new BigInteger("1")` should throw `UnknownFormatConversionException` when previous implementation would throw earlier `MissingFormatWidthException`. That why this check is made later in comparsion to Scala.js
Also ` isNumericConversion(conversion) && flags.zeroPad` condition was added due to missing case which throwns exception in JVM: when using format `%-0e` it worked correctly throwing `MissingFormatWidthException`, but it did not throw in case of `%0e`

Reply:
Could you please file bug reports in the Scala.js repo when Scala.js isn't compliant with the JVM? (unless we already have a bug report for the differences, of course)